pkg/controller/cloudmetrics/cloudmetrics_controller.go

```diff
@@ -17,10 +19,98 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/event"
 	"sigs.k8s.io/controller-runtime/pkg/handler"
 	"sigs.k8s.io/controller-runtime/pkg/manager"
+	customMetrics "sigs.k8s.io/controller-runtime/pkg/metrics"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
+const (
+	postgresFreeStorageAverage = "cro_postgres_free_storage_average"
+	postgresCPUUtilizationAverage = "cro_postgres_cpu_utilization_average"
+	postgresFreeableMemoryAverage = "cro_postgres_freeable_memory_average"
+
+	labelClusterIDKey   = "clusterID"
+	labelResourceIDKey  = "resourceID"
+	labelNamespaceKey   = "namespace"
+	labelInstanceIDKey  = "instanceID"
+	labelProductNameKey = "productName"
+	labelStrategyKey    = "strategy"
+)
+
+// generic list of label keys used for Gauge Vectors
+var labels = []string{
+	labelClusterIDKey,
+	labelResourceIDKey,
+	labelNamespaceKey,
+	labelInstanceIDKey,
+	labelProductNameKey,
+	labelStrategyKey,
+}
+
+// CroGaugeMetric allows for a mapping between an exposed prometheus metric and multiple cloud provider specific metric
+type CroGaugeMetric struct {
+	Name         string
+	GaugeVec     *prometheus.GaugeVec
+	ProviderType map[string]providers.CloudProviderMetricType
+}
+
+// postgresGaugeMetrics stores a mapping between an exposed (postgres) prometheus metric and multiple cloud provider specific metric
+// to add any addition metrics simply add to this mapping and it will be scraped and exposed
+var postgresGaugeMetrics = []CroGaugeMetric{
+	{
+		Name: postgresFreeStorageAverage,
+		GaugeVec: prometheus.NewGaugeVec(
+			prometheus.GaugeOpts{
+				Name: postgresFreeStorageAverage,
+				Help: "The amount of available storage space. Units: Bytes",
+			},
+			labels),
+		ProviderType: map[string]providers.CloudProviderMetricType{
+			providers.AWSDeploymentStrategy: {
+				PromethuesMetricName: postgresFreeStorageAverage,
+				ProviderMetricName:   "FreeStorageSpace",
+				Statistic:            cloudwatch.StatisticAverage,
+			},
+		},
```

Comment:
I'm trying to understand this approach, do you have time to discuss this sometime today?

Reply:
sure, let us know when you are free :+1: